issue-163: added ability to include external files in current.sql.

jnbarlow opened the pull request:
## Description

fixes #163
This PR adds the ability to include external files into current.sql, allowing for better
integration with source control.  Functions, computed columns, etc can live in separate 
diffable files for pull requests, but still take advantage of all the goodness of graphile-migrate
and the `current.sql`.

## Performance impact

If someone decides to include a ton of files in a `current.sql`, the I/O to read these in will make 
things not super pleasant.

## Security impact

unknown... shouldn't be any more risky than using `current.sql`

## Checklist

<!-- If this PR is work in progress, please open it as a "Draft PR". -->
<!-- To tick a checkbox, change it from `[ ]` to `[x]` -->

- [x] My code matches the project's code style and `yarn lint:fix` passes.
- [x] I've added tests for the new feature, and `yarn test` passes.
- [x] I have detailed the new feature in the relevant documentation.
- [ ] I have added this feature to 'Pending' in the `RELEASE_NOTES.md` file (if one exists).
- [ ] If this is a breaking change I've explained why.

<!-- For some Graphile projects the documentation is the README.md file, for
      others please see https://github.com/graphile/graphile.github.io -->

jnbarlow commented:
@benjie Greetings!  If I understood the flow of how sql gets compiled, I think I shimmed my new compile step in so that it should work for everything.  Let me know if I missed something :)

benjie commented:
Amazing! I’ve not reviewed the code yet, but we need it to integrate with watch mode: https://github.com/graphile/migrate/blob/main/src/commands/watch.ts#L217

easiest would be to require that these files all live within one folder (migrations/includes or migrations/library or whatever) and watch that folder, however another option is to dynamically add files to chokidar when they are imported and remove them when they are no longer so. Chokidar has APIs for this.

jnbarlow commented:
@benjie So you're thinking of having a base folder that all the includes live (with Chokidar watching it) so that any time one of them changes it reprocesses the current.sql?  

Is the path something we should make configurable with a default or hard code it as part of the framework?  What about placeholders? Right now the includes are brought in after the placeholder replacement.  We could swap that, but I'm not sure of the use cases would warrant that.  To me, it feels like the includes probably shouldn't have placeholders, but you've got more experience with how people use them.

benjie commented:
Hardcoded is fine to start with; if someone really wants it configurable they can send a PR. Still not sure of the name; perhaps migrations/lib…

Yes, I’m thinking any change to that folder (or descendents) triggers current to run again; we already checksum the result so if it wasn’t relevant we’ll skip it.

It’s very common for roles to be placeholders, particularly on shared hosting (where you grant to a different role on dev versus staging versus production), and in fact that’s what we do in starter via :VISITOR or :VISITOR_ROLE or whatever (I forget the specifics). So it’s essential that the included files are also processed via the placeholder system. I think there’s a handy method you can call on the strings to apply the placeholders, you don’t have to concatenate first if you don’t want to.

benjie commented on __tests__/compile.test.ts:
```diff
@@ -48,3 +53,51 @@ CREATE EXTENSION IF NOT EXISTS pgcrypto WITH SCHEMA public;
+
+it("compiles multiple included files", async () => {
+  mockFs({
+    "foo.sql": "select * from foo;",
+    "bar.sql": "select * from bar;",
+    "baz.sql": "select * from baz;",
```
`--!` include statements should also work inside the fixtures themselves; though we should add a stack of "included files" so we can detect infinite loops.

```suggestion
    "migrations/fixtures/dir1/foo.sql": "select * from foo;",
    "migrations/fixtures/dir2/bar.sql": "select * from bar;",
    "migrations/fixtures/dir3/baz.sql": "--! include dir4/qux.sql",
    "migrations/fixtures/dir4/qux.sql": "select * from qux;",
```

jnbarlow replied:
haha wow. I had thought of maybe doing this, but figured you wouldn't want to support that level of nesting. Let me see what I can do.

benjie commented on __tests__/compile.test.ts:
```diff
@@ -48,3 +53,51 @@ CREATE EXTENSION IF NOT EXISTS pgcrypto WITH SCHEMA public;
+        },
+      },
+      `\
+--!include foo.sql
+--!include bar.sql
+--!include baz.sql
```
```suggestion
--!include dir1/foo.sql
--!include dir2/bar.sql
--!include dir3/baz.sql
```

benjie commented on __tests__/compile.test.ts:
```diff
@@ -48,3 +53,51 @@ CREATE EXTENSION IF NOT EXISTS pgcrypto WITH SCHEMA public;
+  ).toEqual(`\
+select * from foo;
+select * from bar;
+select * from baz;
```
```suggestion
select * from qux;
```

benjie commented on src/commands/run.ts:
```diff
@@ -34,7 +34,9 @@ export async function run<T = any>(
 ): Promise<T[] | undefined> {
   const parsedSettings = await parseSettings(settings, shadow);
-  const sql = compilePlaceholders(parsedSettings, content, shadow);
+  const sql = await compileIncludes(
+    compilePlaceholders(parsedSettings, content, shadow),
+  );
```
This is the wrong time to do this; the code _must_ be included into the committed migration, otherwise as the file evolves newer people who run the migrations from scratch will get a different result.

benjie reviewed (changes requested):
It is essential that these substitutions are made before the SQL is written to `migrations/committed/`, so putting the code in `run` is incorrect - instead it should go in `readCurrentMigration`: https://github.com/graphile/migrate/blob/757194cb2c9cc7a2eebbd8783d0e7065fe504cc9/src/current.ts#L98

Importantly, each file that is read should be subjected to these transforms, even the included files themselves. (Be sure to use a stack to prevent infinite loops.)

For security, please ensure that the file to be imported is actually within the `migrations/fixtures/` folder, e.g. by using `fs.realpath` (on both the resolved path and the fixtures folder) or similar.

Really excited to add this feature finally; it's been on my TODO list for a very long time!

jnbarlow commented:
@benjie Ahh ok, thanks for pointing me to where this should go.  I was trying to figure out the flow of things and shim it into a spot where it would always be called.  Run seemed to be the logical spot that called the placeholder replacement to do that.  Let me work on the updates then

jnbarlow commented on README.md:
```diff
@@ -720,6 +720,47 @@ by `graphile-migrate watch` is defined. By default this is in the
 
+#### Including external files in the current migration
+You can include external files in your `current.sql` to better assist in source
+control. These includes are always relative to the root of where graphile-migrate
+is running.
```
@benjie Do we want to roll with `fixtures` or do we want to use something more generic like `entities` ?  I was asking chatgpt last night if there was a more common term for these sorts of things and `entities` is the best thing it could come up with.

benjie commented:
Yeah calling it in readCurrentMigration means you don't even have to think about placeholders because they're handled for you later. (Placeholders can change at runtime, that's deliberate.)

jnbarlow commented:
@benjie alright, I think I've address everything and pushed.

- hardened the `compileInclude` function to work if the fixtures directory doesn't exist
- moved the `compileInclude` call to `readCurrentMigration`
- fixed all relevant tests after including this step
- updated the watch to include the fixtures directory

Give this a look whenever you get a chance.  The only thing I've not done yet is actually run the code (been relying on tests to check everything).   I'm not super sure how to go about compiling it down to something I can execute.

benjie commented:
Awesome; will take a look when I'm back from the winter break.

> I'm not super sure how to go about compiling it down to something I can execute.

You should be able to run `yarn prepack` which will build everything. Then you can run `yarn link` to create a global version of it, and in a clean folder somewhere (or in your project that uses migrate) you can do `yarn link graphile-worker` to install the version. I'm not 100% sure this will work, it sometimes gets confused due to modules coming from multiple places, but migrate is simple enough that I think it will.

Alternatively you can run `yarn pack` which will create a tgz containing the module, and then you can install that tgz into a project somewhere via `yarn add file:/path/to/local/tarball.tgz` or equivalent in whatever package manager you use.

benjie commented on src/current.ts:
```diff
@@ -178,7 +182,8 @@ export async function readCurrentMigration(
     }
-    return wholeBody;
+
+    return await compileIncludes(_parsedSettings, wholeBody);
```
`return await` is unnecessary (and consumes an extra tick) unless you're inside a `try{}` block.

```suggestion
    return compileIncludes(_parsedSettings, wholeBody);
```

benjie commented on src/current.ts:
```diff
@@ -103,7 +107,7 @@ export async function readCurrentMigration(
 
     // If file doesn't exist, treat it as if it were empty.
-    return content || "";
+    return await compileIncludes(_parsedSettings, content || "");
```
```suggestion
    return compileIncludes(_parsedSettings, content || "");
```

jnbarlow commented:
@benjie another round pushed up

- updated watch to proper `**/*` to get everything in the `fixtures` directory
- had to add regex escaping to imported files... files with `$` was breaking in the replacement
  -  Can you think of any other SQL allowed characters that would need to be escaped?
- added tests

I was able to get it to run locally too, but ran into a problem where `yarn add` kept importing old versions.   I manually extracted the built file and it worked.

benjie commented on src/migration.ts:
```diff
@@ -122,6 +122,75 @@ export function compilePlaceholders(
+  } else {
+    return compiledContent;
+  }
+}
```
I started refactoring this, but turns out we're quite limited by our out of date dependencies/Node.js version/etc. So I've raised #198 to upgrade the codebase to use the latest of everything. I'll merge that in here shortly.

benjie reviewed (changes requested):
Thanks so much for taking this on, sorry it's taken me a while to get around to, January has been hectic!

I'm going to take on getting this over the finish line, thanks again!

jnbarlow commented on src/commands/watch.ts:
```diff
@@ -214,29 +214,35 @@ export async function _watch(
+    const watcher = chokidar.watch(
+      [
+        currentLocation.path,
+        `${parsedSettings.migrationsFolder}/fixtures/**/*`,
```
@benjie Does it pick up subfolders without it? For watch to pick up changes in subfolders, I had to add the /*/**. (granted, I started with /*). We want to be able to watch things like `fixtures/functions/foo.sql`

benjie replied:
As far as I know, folders are watched recursively. This issues pretty old but says so:

https://github.com/paulmillr/chokidar/issues/67#issuecomment-28514857

jnbarlow replied:
ahh ok, so maybe me putting it in as `/*` broke it the first time.  Carry on :)

jnbarlow commented on src/migration.ts:
```diff
@@ -122,6 +122,75 @@ export function compilePlaceholders(
+export async function compileIncludes(
+  parsedSettings: ParsedSettings,
+  content: string,
+  processedFiles: Array<string> = [],
```
Haha sorry, TS isn't my strongsuit... I always default to OG ways

benjie commented:
@jnbarlow I think this is good to merge now; would you care to give it a test and make sure it still works well (in particular, ensure that watch mode functions correctly)?

jnbarlow commented:
@benjie Sure!.  I'll give that a test tonight and let you know

jnbarlow commented:
@benjie well, watch works, but the migrations aren't running.  I'm getting this error: 

`Error occurred whilst processing migration: Cannot read properties of undefined (reading 'bold')`

jnbarlow commented:
@benjie everything works :) ...  Ignore my last message if you saw it, I was doing something boneheaded when I tested it :)
